man/man8/os-indications.8
.\" Copyright © 2019-2021 Johnothan King. All rights reserved.
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to deal
.\" in the Software without restriction, including without limitation the rights
.\" to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in all
.\" copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\" OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
.\" SOFTWARE.
.\"
.Dd May 22, 2020
.Dt OS-INDICATIONS 8
.Os
.Sh NAME
.Nm os-indications
.Nd Set the OsIndications UEFI variable
for booting into firmware setup
.Sh SYNOPSIS
.Nm
.Op Fl qu?
.Sh DESCRIPTION
This program controls whether or not the UEFI variable
.Nm OsIndications
is set to boot into
the system firmware's UI on the system's next boot.
When run without any flags or options,
.Nm
will set
.Nm OsIndications
to boot into the system firmware on the next boot.

This program accepts the following flags:

.Nm -q, --quiet
Disable
.Nm
output (unless an error occurs).

.Nm -u, --unset
Unset the 0x0000000000000001 value in
.Nm OsIndications
to reverse any previous actions that were taken.

.Nm -?, --help
Show
.Nm
usage information.
#DEF Linux
.Sh FILES
.Em /sys/firmware/efi/efivars
This is the efivarfs that contains the system's UEFI variables.
Non-standard variables are made immutable by default.

.Em /sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c
The
.Nm OsIndications
UEFI variable that
.Nm
writes to.
If this file is immutable,
.Nm
will fail to set the OsIndications UEFI variable.
This can be fixed by using
.Nm chattr(1) .

.Em /sys/firmware/efi/efivars/OsIndicationsSupported-8be4df61-93ca-11d2-aa0d-00e098032b8c
This is the UEFI variable that tells
.Nm
if the system supports rebooting into the firmware's interface from the current OS.
#ENDEF
.Sh SEE ALSO
efivar(1)
.Sh AUTHOR
Johnothan King


CHANGELOG.md
**NOTICE:** LeanInit v2 is **not** backward compatible with v1. To upgrade from v1 to v2, run `make uninstall` on the v1 branch then run `make install` on the v2 (master) branch.

**v2.0.0 Release Candidate 1 Changelog:**
* Heavily optimized various parts of LeanInit to improve performance and decrease its footprint.
    * LeanInit's LDFLAGS are now the following: `-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now`.
    * LeanInit's ELF binaries are now stripped of `.comment`, `.gnu-version`, `.GCC.command.line`, `.note.gnu.gold-version` and with `--strip-unneeded` after being built to decrease file size.
    * ZFS file system mounting is now faster.
    * The `zloop()` thread and system shutdown are now both faster as the `zstatus` variable has been replaced with usage of waitpid(3).
    * Services that are not compatible with the target OS will no longer be installed with `make install`.
    * `leaninit-service --status-all` is now faster.
    * Various parts of rc.svc(8) are now faster.
    * Removed the useless usage of `return` from the LeanInit signal handler and stall.
    * The current_signal variable will now always be reset at the end of the main `for(;;)` loop.
    * `unsigned int` is now used in place of `size_t`.
    * More strings are globbed to increase stability and performance.
    * Globs are now used in place of executing ls(1) when desirable.
    * Archaic usage of `\` has been removed from rc.svc(8).
    * The .gitignore file is now a little bit shorter.
* Added support for standard getty implementations such as agetty(8) and the FreeBSD getty(8).
    * LGetty as a result has been removed from the main LeanInit source tree.
    * The format for `/etc/leaninit/ttys` has been drastically changed. The LeanInit ttys(5) file is now styled after inittab(5).
* Ported os-indications(8) to FreeBSD using the libefivar API.
* Moved the services folder from `/etc/leaninit/svc.d` to `/etc/leaninit/svc`.
* The consolekit and fstrim services have been removed.
* run() is now called automatically when rc.svc(8) is sourced if the environment variable `$__SVC` is not set to false.
* Reimplemented the old fork() function from pre-1.0 commits with the purpose of writing the PIDs of forked commands.
* Support for `--firmware-setup` has been added to halt(8).
* os-indications(8) now supports the `--quiet` flag, which disables output.
* Added support for a new function called restart() to LeanInit services. This function will be run instead of main() when a service is restarted (if this function exists).
* The naming format for programs and man pages has been changed from lPROGNAME to leaninit-PROGNAME (e.g. lservice -> leaninit-service).
* The debug programs have been moved to `./debug`.
* Added a debug program to `./debug` called `signal-interfere`. This tool will attempt to override the first signal sent to init(8) by quickly sending another signal to init before the signal-interfere daemon is closed. Sending `SIGALRM` can cause certain init systems to prematurely shutdown when using timers based on alarm(2). Sending `SIGTERM` after sending a different signal (such as `SIGINT`) can override the target runlevel and put init into the wrong state.
* Improved the error checking in the stall debug program and gave it colored output.
* Stall will no longer block the signal `SIGINT`.
* Moved the universal include file to the include folder, renamed it to `leaninit.h` and made the debug programs use this header.
* The elogind service now has more comments to improve readability.
* Replaced the erroneous 'NetworkManager' comment in the wicd service with a 'Wicd' comment.
* Removed support for runlevel variables as they were useless and caused race conditions.
    * The command lrunlevel(8) has consequently been removed as well.
* Removed support for `/var/log/leaninit.log.2` (as well as `/var/log/leaninit.log.old.2`) due to security concerns.
* Removed outdated information from the `README.md` file.
* Fixed a bug that caused the wpa_supplicant service to fail to restart on FreeBSD.
* Fixed a bug that caused os-indications(8) to display the wrong name in verbose output if it was renamed to something else.
* Usage information for services is now accurate even if the service is not located in `/etc/leaninit/svc` and/or if no arguments are given.
* `DEF` and `#ifdef` ordering across LeanInit is now more consistent.
* The default rc.conf(5) file has been slightly changed.
* Changed the message `Launching sysctl(8)...` to `Executing sysctl(8)...` in rc(8).
* Spaces are now used instead of tabs to improve formatting.
* Improved the formatting of `init.c` specifically.
* Made various improvements to the man pages:
    * Added information about `$__svcname` and `$__svcpid` to the leaninit-rc.svc(8) man page.
    * Changed 'path' to 'default path' in the leaninit-rc.svc(8) man page.
    * Improved the formatting of the leaninit-rc.conf(5) and leaninit-rc.svc(8) man pages.
    * Fixed some grammar in the os-indications(8) man page.
    * Changed the self reference of `halt(8)` to `halt` in the leaninit-halt(8) man page.
    * Updated the dates in every single man page and in the debug tools.


cmd/os-indications.c
/*
 * Copyright © 2019-2021 Johnothan King. All rights reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/*
 * os-indications -- Set OsIndications for booting into firmware setup.
 *
 * The libefivar API is used when compiled on FreeBSD, while the efivarfs and
 * C Standard APIs are used on Linux to avoid requiring an extra dependency.
 *
 * This small tool only supports Linux and FreeBSD at this point in time.
 */

#include <leaninit.h>

int main(int argc, char *argv[])
{
#if !defined(__linux__)
    // GUID for OsIndications
    efi_guid_t global_guid = EFI_GLOBAL_GUID;
#endif

    // Boolean variables and long options
    bool verbose = true;
    bool unset = false;
    struct option long_options[] = { { "quiet", no_argument, NULL, 'q' },
                                     { "unset", no_argument, NULL, 'u' },
                                     { "help", no_argument, NULL, '?' },
                                     { NULL, 0, NULL, 0 } };

    // Parse options
    int args;
    while ((args = getopt_long(argc, argv, "qu?", long_options, NULL)) != -1)
        switch (args) {

            // Display usage info
            case '?':
                printf("Usage: %s [-qu?]\n"
                       "  -q, --quiet   Disable output (unless there was an error)\n"
                       "  -u, --unset   Revert changes made by os-indications\n"
                       "  -?, --help    Show this usage information\n",
                       __progname);
                return 1;

            // Quiet mode
            case 'q':
                verbose = false;
                break;

            // Unset OsIndications
            case 'u':
                unset = true;
                break;
        }

    // Error checks
    if unlikely (getuid() != 0) {
        printf(RED "* Permission denied!" RESET "\n");
        return 1;
#if !defined(__linux__)
    } else if very_unlikely (efi_variables_supported() == 0) {
#else
    } else if very_unlikely (
        access("/sys/firmware/efi/efivars/OsIndicationsSupported-8be4df61-93ca-11d2-aa0d-00e098032b8c", R_OK) != 0) {
#endif
        printf(RED "* This system does not support OsIndications!" RESET "\n");
        return 1;
    }

#if defined(__linux__)
    // Write efi_data (Linux efivarfs API)
    if likely (!unset) {
        FILE *fd = fopen("/sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c", "w+");
        unsigned int efi_data[2] = { 0x07, 0x01 };
        if unlikely (fwrite(efi_data, 2, 3, fd) == 0) {
            fclose(fd);
            printf(RED "* Failed to write the changes to OsIndications!" RESET "\n");
            return 1;
        }
        fclose(fd);

        // Delete OsIndications to unset it
    } else
        unlink("/sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c");

#else
    // Set OsIndications for booting into firmware setup (FreeBSD libefivar API)
    unsigned char efi_boot = 0x01;
    if likely (!unset)
        efi_set_variable(global_guid, "OsIndications", &efi_boot, 1, 0x07);

    // Delete OsIndications to unset it
    else
        efi_del_variable(global_guid, "OsIndications");
#endif

    // Notify the user of the change if --quiet was not passed
    if (verbose) {
        if likely (!unset) {
            printf(CYAN "* " WHITE "This system will now boot into the firmware's UI the next time it boots.\n" CYAN
                        "* " WHITE "Run `%s --unset` to revert this change." RESET "\n",
                   __progname);
        } else {
            printf(CYAN "* " WHITE "This system will NOT boot into the firmware's UI the next time it boots.\n" CYAN
                        "* " WHITE "Any prior changes made by %s have been reverted." RESET "\n",
                   __progname);
        }
    }

    return 0;
}
